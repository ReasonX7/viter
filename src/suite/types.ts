import { BenchResult } from "@/runners/types";

export type SuiteItemStatus = "run" | "skip" | "todo";

export type BenchResultStatus = "pending" | "error" | "done";

export type IdObject = {
  id: string;
};

export type StatusObject<S> = {
  status: S;
};

export type DescriptionObject = {
  description: string;
};

export type DescriptionItem = IdObject &
  DescriptionObject &
  StatusObject<SuiteItemStatus> & {
    children: string[];
  };

export type BenchGroup = IdObject & {
  benchFunctionIds: string[];
};

export type BenchResultItem = IdObject &
  StatusObject<BenchResultStatus> & {
    data: bigint[] | null;
    error: Error | null;
  };
