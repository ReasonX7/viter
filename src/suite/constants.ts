export const BenchResultEvent = {
  Done: "BenchResult:Done",
  Error: "BenchResult:Error",
} as const;
