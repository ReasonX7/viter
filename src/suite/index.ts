import { default as EventEmitter } from "node:events";
import type { BenchFunction, BenchResult, RunBenchmarkFunc, RunOptions } from "@/runners/types";
import { BenchResultEvent } from "@/suite/constants";
import type { BenchGroup, BenchResultItem, DescriptionItem, SuiteItemStatus } from "@/suite/types";

export const descriptionStore = new Map<string, DescriptionItem>();
export const benchFunctionStore = new Map<string, BenchFunction>();
export const benchGroupStore = new Map<string, BenchGroup>();
export const benchResultStore = new Map<string, BenchResultItem>();

export const events = new EventEmitter();

const descriptionIdStack: string[] = [];

export const resetAllStores = () => {
  descriptionStore.clear();
  benchFunctionStore.clear();
  benchGroupStore.clear();
  benchResultStore.clear();
};

export const createPendingBenchResult = (id: string) =>
  ({
    id,
    status: "pending",
    data: null,
    error: null,
  }) satisfies BenchResultItem;

export const createDoneBenchResult = (id: string, data: BenchResult) =>
  ({
    id,
    status: "done",
    data,
    error: null,
  }) satisfies BenchResultItem;

export const createErrorBenchResult = (id: string, error: Error) =>
  ({
    id,
    status: "error",
    data: null,
    error,
  }) satisfies BenchResultItem;

const createId = (...strings: (string | undefined)[]) => strings.filter(Boolean).join(" >> ");

const $describe = (description: string, status: SuiteItemStatus, func: () => void) => {
  const parentId = descriptionIdStack.at(-1);
  const descId = createId(parentId, description);

  if (parentId) {
    // biome-ignore lint/style/noNonNullAssertion: <explanation>
    descriptionStore.get(parentId)!.children.push(descId);
  }

  descriptionStore.set(descId, { id: descId, status, description, children: [] });
  descriptionIdStack.push(descId);
  func();
  descriptionIdStack.pop();
};

const $bench = (description: string, status: SuiteItemStatus, func: BenchFunction) => {
  const descId = descriptionIdStack.at(-1);
  const funcId = createId(descId, description);

  if (descId == null) {
    throw new Error("[viter]: `bench` function must be inside `description`");
  }

  const group = benchGroupStore.get(descId) || { id: descId, benchFunctionIds: [] };

  group.benchFunctionIds.push(funcId);
  benchGroupStore.set(descId, group);
  benchResultStore.set(descId, createPendingBenchResult(descId));
  benchFunctionStore.set(funcId, func);

  $describe(description, status, () => {});
};

export const describe = (name: string, fn: () => void) => {
  $describe(name, "run", fn);
};

describe.skip = (name: string, fn: () => void) => {
  $describe(name, "skip", fn);
};

describe.todo = (name: string, fn: () => void = () => {}) => {
  $describe(name, "todo", fn);
};

export const bench = (name: string, fn: () => void) => {
  $bench(name, "run", fn);
};

bench.skip = (name: string, fn: () => void) => {
  $bench(name, "skip", fn);
};

bench.todo = (name: string, fn: () => void = () => {}) => {
  $bench(name, "todo", fn);
};

export const start = async (runBenchmark: RunBenchmarkFunc, runOptions: RunOptions) => {
  for (const [descId, group] of benchGroupStore) {
    // biome-ignore lint/style/noNonNullAssertion: <explanation>
    const funcs = group.benchFunctionIds.map((id) => benchFunctionStore.get(id)!);

    await runBenchmark(funcs, runOptions)
      .then((result) => {
        benchResultStore.set(descId, createDoneBenchResult(descId, result));
        events.emit(BenchResultEvent.Done, descId);
      })
      .catch((error) => {
        benchResultStore.set(descId, createErrorBenchResult(descId, error));
        events.emit(BenchResultEvent.Error, descId);
      });
  }
};

// export const formConsoleOutput = () => {
//   const descIds = Array.from(descriptionStore).map((value) => value[0]);
//   return descIds.map((id) => )
// };
