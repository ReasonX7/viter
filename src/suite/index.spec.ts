import type { BenchResult, RunBenchmarkFunc } from "@/runners/types";
import {
  events,
  bench as $bench,
  describe as $describe,
  benchFunctionStore,
  benchGroupStore,
  benchResultStore,
  descriptionStore,
  resetAllStores,
  start,
} from "@/suite";
import { BenchResultEvent } from "@/suite/constants";
import { afterEach, describe, expect, it, vi } from "@reasonx/tools/vitest";
// import type { DescribeItem, BenchItem } from "@/suite";

describe("suite", () => {
  afterEach(() => {
    resetAllStores();
  });

  it("should create & run the successful benchmark", async () => {
    /**
     * Setup benchmarks suite.
     */
    const benchPush = vi.fn();
    const benchPop = vi.fn();
    const benchDeepWalk = vi.fn();

    const handleDoneBenchResult = vi.fn();
    const handleErrorBenchResult = vi.fn();

    // TODO: Add "todo" & "skip" describes and benches.
    // TODO: Add bench function that throws an error.
    // TODO: Add bench function that never resolves and has to be stopped by an AbortSignal.
    $describe("Utils", () => {
      $describe("LinkedList", () => {
        $bench("LinkedList Push", benchPush);
        $bench("LinkedList Pop", benchPop);
      });

      $describe("Object", () => {
        $bench("Deep Walk", benchDeepWalk);
      });
    });

    expect(descriptionStore).toMatchSnapshot();
    expect(benchFunctionStore).toMatchSnapshot();
    expect(benchGroupStore).toMatchSnapshot();

    /**
     * Run benchmark suite.
     */
    events.on(BenchResultEvent.Done, handleDoneBenchResult);
    events.on(BenchResultEvent.Error, handleDoneBenchResult);

    const runBenchmarkResults: BenchResult[] = [
      // LinkedList suite results:
      {
        calcInfo: {
          testTime: 1000,
          iterations: 900,
          iterationCycles: 100,
          funcsCount: 2,
        },
        cycles: [20n, 30n],
        absoluteDiff: [],
        relativeDiff: undefined,
      },

      // Object suite results:
      {
        calcInfo: {
          testTime: 1000,
          iterations: 500,
          iterationCycles: 10,
          funcsCount: 1,
        },
        cycles: [10n],
        absoluteDiff: [],
        relativeDiff: undefined,
      },
    ];
    const runBenchmarkMock: RunBenchmarkFunc = vi.fn().mockImplementation((funcs) => {
      if (funcs.length === 2) {
        // This is for "LinkedList" suite.
        return Promise.resolve(runBenchmarkResults[0]);
      }

      // This is for "Object" suite.
      return Promise.resolve(runBenchmarkResults[1]);
    });

    await start(runBenchmarkMock, { warmupTime: 2000, runTime: 1000 });

    expect(runBenchmarkMock).toHaveBeenCalledTimes(2);
    expect(runBenchmarkMock).toHaveBeenNthCalledWith(1, [benchPush, benchPop], {
      warmupTime: 2000,
      runTime: 1000,
    });
    expect(runBenchmarkMock).toHaveBeenNthCalledWith(2, [benchDeepWalk], {
      warmupTime: 2000,
      runTime: 1000,
    });

    expect(handleErrorBenchResult).toHaveBeenCalledTimes(0);
    expect(handleDoneBenchResult).toHaveBeenCalledTimes(2);
    expect(handleDoneBenchResult).toHaveBeenNthCalledWith(1, "Utils >> LinkedList");
    expect(handleDoneBenchResult).toHaveBeenNthCalledWith(2, "Utils >> Object");

    console.log("Description:");
    console.log(descriptionStore);
    console.log("Description:");
    console.log(descriptionStore);
    console.log("Bench results:");
    console.log(benchResultStore);
  });
});
