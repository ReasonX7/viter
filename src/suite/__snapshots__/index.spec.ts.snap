// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`suite > should create & run the successful benchmark 1`] = `
Map {
  "Utils" => {
    "children": [
      "Utils >> LinkedList",
      "Utils >> Object",
    ],
    "description": "Utils",
    "id": "Utils",
    "status": "run",
  },
  "Utils >> LinkedList" => {
    "children": [
      "Utils >> LinkedList >> LinkedList Push",
      "Utils >> LinkedList >> LinkedList Pop",
    ],
    "description": "LinkedList",
    "id": "Utils >> LinkedList",
    "status": "run",
  },
  "Utils >> LinkedList >> LinkedList Push" => {
    "children": [],
    "description": "LinkedList Push",
    "id": "Utils >> LinkedList >> LinkedList Push",
    "status": "run",
  },
  "Utils >> LinkedList >> LinkedList Pop" => {
    "children": [],
    "description": "LinkedList Pop",
    "id": "Utils >> LinkedList >> LinkedList Pop",
    "status": "run",
  },
  "Utils >> Object" => {
    "children": [
      "Utils >> Object >> Deep Walk",
    ],
    "description": "Object",
    "id": "Utils >> Object",
    "status": "run",
  },
  "Utils >> Object >> Deep Walk" => {
    "children": [],
    "description": "Deep Walk",
    "id": "Utils >> Object >> Deep Walk",
    "status": "run",
  },
}
`;

exports[`suite > should create & run the successful benchmark 2`] = `
Map {
  "Utils >> LinkedList >> LinkedList Push" => [MockFunction spy],
  "Utils >> LinkedList >> LinkedList Pop" => [MockFunction spy],
  "Utils >> Object >> Deep Walk" => [MockFunction spy],
}
`;

exports[`suite > should create & run the successful benchmark 3`] = `
Map {
  "Utils >> LinkedList" => {
    "benchFunctionIds": [
      "Utils >> LinkedList >> LinkedList Push",
      "Utils >> LinkedList >> LinkedList Pop",
    ],
    "id": "Utils >> LinkedList",
  },
  "Utils >> Object" => {
    "benchFunctionIds": [
      "Utils >> Object >> Deep Walk",
    ],
    "id": "Utils >> Object",
  },
}
`;
